File: src/gui/components/SliderSetting.java
package gui.components;

import javafx.beans.property.DoubleProperty;
import javafx.scene.control.Label;
import javafx.scene.control.Slider;
import javafx.scene.control.TextField;

import java.util.Optional;

public class SliderSetting
{
    private final boolean allowDecimalNumbers;
    private Optional<Label> label;
    private Optional<NumberTextField> numberTextField;
    private Optional<Slider> slider;

    public SliderSetting(boolean allowDecimalNumbers)
    {
        this.allowDecimalNumbers = allowDecimalNumbers;
        label = Optional.empty();
        numberTextField = Optional.empty();
        slider = Optional.empty();
    }

    public DoubleProperty valueProperty()
    {
        return getSlider().valueProperty();
    }

    public Label getLabel()
    {
        return label.orElseGet(
            () -> {
                label = Optional.of(new Label());
                return label.get();
            });
    }

    public TextField getNumberTextField()
    {
        return numberTextField.orElseGet(() ->
            {
                numberTextField = Optional.of(new NumberTextField());
                numberTextField.get().textProperty().addListener((value,oldValue,newValue) ->
                    {
                        getSlider().setValue(Double.parseDouble(newValue));
                        getNumberTextField().setText(allowDecimalNumbers
                            ? Double.toString(getSlider().getValue())
                            : Integer.toString((int) Math.round(getSlider().getValue())));
                    });
                return numberTextField.get();
            });
    }

    public Slider getSlider()
    {
        return slider.orElseGet(() ->
            {
                slider = Optional.of(new Slider());
                slider.get().setMaxWidth(Double.MAX_VALUE);
                slider.get().valueProperty().addListener((value,oldValue,newValue) ->
                    {
                        getNumberTextField().setText(newValue.toString());
                    });
                return slider.get();
            });
    }
}

File: netsim/src/gui/components/SliderSetting.kt
package gui.components

import javafx.beans.property.DoubleProperty
import javafx.scene.control.Label
import javafx.scene.control.Slider
import parseDouble

class SliderSetting(private val allowDecimalNumbers:Boolean)
{
    val label:Label = Label()
    var numberTextField:NumberTextField = NumberTextField()
    var slider:Slider = Slider()

    init
    {
        // configure numberTextField so when its value changes, it updates slider
        numberTextField.textProperty().addListener(
            {
                value,oldValue,newValue ->
                slider.value = parseDouble(newValue)
                numberTextField.text = if (allowDecimalNumbers)
                    slider.value.toString()
                    else Math.round(slider.value).toString()
            })

        // configure slider so when its value changes, it updates numberTextField
        slider.maxWidth = java.lang.Double.MAX_VALUE
        slider.valueProperty().addListener(
            {
                value,oldValue,newValue ->
                numberTextField.text = newValue.toString()
            })
    }

    fun valueProperty():DoubleProperty
    {
        return slider.valueProperty()
    }
}
